### PR TITLE
ROC-7557: Issue general letter to claimant or defendant through CCD UI

### PR DESCRIPTION
**Pull request by akritis22**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-7557

### Change description ###
A caseworker can issue a letter to a party in an online civil money claim

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -95,9 +95,10 @@ bankHolidays:
     url: "https://www.gov.uk"
 
 doc_assembly:
-  url: ${DOC_ASSEMBLY_URL:false}
+  url: ${DOC_ASSEMBLY_URL:true}
```
Comment (dharmendrak) [inline src/main/resources/application.yml]: please revert back to false

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(caseDetailsConverter.convertToMap(updatedCase))
+                .build();
+
+        } catch (Exception e) {
+            logger.info("General Letter printing and case documents update failed", e);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]: We're not sure nothing has been sent in this case - the exception could have been thrown afterwards.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java [hunk 3]
```diff
@@ -94,4 +94,10 @@
     private boolean expertReportPermissionPartyGivenToDefendant;
     private List<CCDCollectionElement<String>> expertReportInstructionClaimant;
     private List<CCDCollectionElement<String>> expertReportInstructionDefendant;
+    private String partyName;
+    private CCDAddress partyAddress;
+    private String body;
+    private String caseName;
+    private String caseWorkerName;
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java]: Capitalisation for caseworker is inconsistent with codebase
```suggestion
    private String caseworkerName;
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyServiceTest.java [hunk 4]
```diff
@@ -59,28 +63,23 @@ public void setup() {
             userService,
             LEGAL_ADVISOR_TEMPLATE_ID,
             JUDGE_TEMPLATE_ID);
-
+        ccdCase = SampleData.addCCDOrderGenerationData(ccdCase);
         when(userService.getUserDetails(eq(BEARER_TOKEN))).thenReturn(JUDGE);
+        docAssemblyResponse = Mockito.mock(DocAssemblyResponse.class);
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyServiceTest.java]: docAssemblyResponse has an @Mock notation. This isn't needed.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            .build();
+    }
+
+    private String getDocumentNumber(CCDCase ccdCase) {
+        String number = String.valueOf((ccdCase.getCaseDocuments()
+            .stream().filter(c ->
+                c.getValue().getDocumentType()
+                    .equals(GENERAL_LETTER) && c.getValue().getDocumentName()
+                    .contains(LocalDate.now().toString())
+            ).collect(Collectors.toList()).size() + 1));
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]: Try to keep stream operations as simple and atomic as possible:
```suggestion
            .stream()
            .filter(c -> c.getValue().getDocumentType().equals(GENERAL_LETTER))
            .filter(c -> c.getValue().getDocumentName().contains(LocalDate.now().toString()))
            .collect(Collectors.toList()).size() + 1));
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyService.java [hunk 6]
```diff
@@ -68,12 +68,31 @@ public DocAssemblyResponse createOrder(CCDCase ccdCase, String authorisation) {
         );
     }
 
+    public DocAssemblyResponse createGeneralLetter(CCDCase ccdCase, String authorisation, String templateId) {
+        UserDetails userDetails = userService.getUserDetails(authorisation);
+
+        logger.info("Doc assembly service: creating request for doc assembly");
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyService.java]: Logging could be more descriptive
```suggestion
        logger.info("Doc assembly service: creating general letter request for doc assembly");
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyService.java [hunk 7]
```diff
@@ -68,12 +68,31 @@ public DocAssemblyResponse createOrder(CCDCase ccdCase, String authorisation) {
...
+
+        DocAssemblyRequest docAssemblyRequest = DocAssemblyRequest.builder()
+            .templateId(templateId)
+            .outputType(OutputType.PDF)
+            .formPayload(docAssemblyTemplateBodyMapper.generalLetterBody(ccdCase, userDetails))
+            .build();
+
+        logger.info("Doc assembly service: sending request to doc assembly");
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/DocAssemblyService.java]: Logging could be more descriptive
```suggestion
        logger.info("Doc assembly service: sending general letter request to doc assembly");
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            .add(claimDocument)
+            .build();
+    }
+
+    private String getDocumentNumber(CCDCase ccdCase) {
+        String number = String.valueOf((ccdCase.getCaseDocuments()
+            .stream().filter(c ->
+                c.getValue().getDocumentType()
+                    .equals(GENERAL_LETTER) && c.getValue().getDocumentName()
+                    .contains(LocalDate.now().toString())
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]:
```suggestion
                c.getValue().getDocumentType().equals(GENERAL_LETTER) 
                && c.getValue().getDocumentName().contains(LocalDate.now().toString())
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java (hunk 5 above)
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]: can be simplified
```suggestion
            ).count() + 1));
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java (hunk 3 above)
Comment (akritis22) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java]: this is for the time being as to match the new template as this field is caseWorkerName in the template

Comment (daveoh) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java]: We can update the template though.

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java (hunk 2 above)
Comment (akritis22) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]: discussed this with Chris and now only catching exception when fails on docmosis call or bulkprint.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java (hunk 3 above)
Comment (akritis22) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBody.java]: yes @dharmendrak is working on that..

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java [hunk 9]
```diff
@@ -101,4 +102,23 @@ public DocAssemblyTemplateBody from(CCDCase ccdCase, UserDetails userDetails) {
...
     }
+
+    public DocAssemblyTemplateBody generalLetterBody(CCDCase ccdCase, UserDetails userDetails) {
+        LocalDate currentDate = LocalDate.now(clock.withZone(UTC_ZONE));
+        return DocAssemblyTemplateBody.builder()
+            .currentDate(currentDate)
+            .referenceNumber(ccdCase.getPreviousServiceCaseReference())
+            .caseworkerName(userDetails.getFullName())
+            .caseName(ccdCase.getCaseName())
+            .partyName(ccdCase.getIssueLetterContact() == CCDContactPartyType.CLAIMANT
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java]: Determining the party to be done first, so the following 10 statements can be simplified

**Comment by karthikkannankainos** [comment]
is the issuing letter only one time, or caseworker can issue multiple letters?

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDCase.java [hunk 10]
```diff
@@ -115,6 +115,7 @@
     private CCDHearingDurationType estimatedHearingDuration;
 
     private CCDDocument draftOrderDoc;
+    private CCDDocument draftLetterDoc;
```
Comment (karthikkannankainos) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDCase.java]: if the letter can be issued multiple times, we cant have it at the CCDCase level

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java [hunk 11]
```diff
@@ -140,9 +141,34 @@ public void printPdf(Claim claim, List<Printable> documents) {
         );
     }
 
+    @Override
+    public void printGeneralLetterPdf(Claim claim, List<Printable> documents) {
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java]: printGeneralLetterPdf is same implementation of printLetterPdf except for the wrapGeneralLetterInMap. its better to have it as same method and just creating the map of parameters to be done either as supplier or function.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java [hunk 12]
```diff
@@ -166,4 +192,12 @@ private String readDocuments(Document document) {
         additionalData.put(ADDITIONAL_DATA_CASE_REFERENCE_NUMBER_KEY, claim.getReferenceNumber());
         return additionalData;
     }
+
+    private Map<String, Object> wrapGeneralLetterInMap(Claim claim) {
+        Map<String, Object> additionalData = new HashMap<>();
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java]: Only the letter type is different. So all the methods can be organized to have just a type as parameter.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/PrintService.java [hunk 13]
```diff
@@ -9,4 +9,6 @@
     void print(Claim claim, List<Printable> documents);
 
     void printPdf(Claim claim, List<Printable> documents);
+
+    void printGeneralLetterPdf(Claim claim, List<Printable> documents);
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/PrintService.java]: we can use the same printPdf by passing additional letter type as parameter.

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/documents/BulkPrintService.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5201799&bid=17062678)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+            .addAll(ccdCase.getCaseDocuments())
+            .add(claimDocument)
+            .build();
+    }
+
+    private String getDocumentNumber(CCDCase ccdCase) {
+        String number = String.valueOf((ccdCase.getCaseDocuments()
+            .stream()
+            .filter(c -> c.getValue().getDocumentType().equals(GENERAL_LETTER))
+            .filter(c -> c.getValue().getDocumentName().contains(LocalDate.now().toString()))
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java]: Map to the values so you don't need to go through them each time on the filters.
```suggestion
            .stream()
            .map(CCDCollectionElement::getValue)
            .filter(c -> c.getDocumentType().equals(GENERAL_LETTER))
            .filter(c -> c.getDocumentName().contains(LocalDate.now().toString()))
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java [hunk 15]
```diff
@@ -101,4 +105,35 @@ public DocAssemblyTemplateBody from(CCDCase ccdCase, UserDetails userDetails) {
...
+            .caseWorkerName(ccdCase.getCaseworkerName())
+            .caseName(ccdCase.getCaseName())
+            .partyName(partyName)
+            .partyAddress(partyAddress)
+            .body(ccdCase.getLetterContent())
+            .build();
+    }
+
+    private CCDAddress getDefendantAddress(CCDRespondent respondent) {
+        return Optional.ofNullable(respondent.getPartyDetail()).isPresent()
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java]: No point wrapping in an Optional just to check for presence.
```suggestion
        return respondent.getPartyDetail() != null
```

**Pull request merged**